fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/planning/BooleanNormalizer.java

```diff
@@ -39,46 +41,138 @@
 /**
  * A normalizer of a tree of {@link QueryComponent} predicates into disjunctive normal form.
  * <p>
- * The full normal form has a single {@code Or} at the root, all of whose children at {@code And}, none of whose children have other Boolean operators.
- * This is abbreviated to exclude parent Boolean nodes with only one child.
+ * The full normal form has a single {@code Or} at the root, all of whose children at {@code And}, none of whose children
+ * have other Boolean operators. This is abbreviated to exclude parent Boolean nodes with only one child.
  * The intermediate form for the normalizer is a list of lists.
  * </p>
+ *
+ * <p>
+ * The <em>size</em> of a Boolean expression in disjunctive normal form (DNF) is the number of terms in the outermost
+ * {@code Or} <a href="http://www.contrib.andrew.cmu.edu/~ryanod/?p=646">[O'Donnell 2014]</a>.
+ * The {@code BooleanNormalizer} will not normalize a {@link QueryComponent} if the normalized form would have a size
+ * that exceeds the size limit. This limit is useful to avoid wastefully normalizing expressions with a very large DNF.
+ * In some cases, such as a large expression in conjunctive normal form (CNF), attempting to normalize such an expression
+ * will cause out-of-memory errors.
+ * </p>
  */
 @API(API.Status.INTERNAL)
 public class BooleanNormalizer {
-    private BooleanNormalizer() {
+    /**
+     * The default limit on the size of the DNF that will be produced by the normalizer.
+     */
+    public static final int DEFAULT_SIZE_LIMIT = 1_000_000;
+    private static final BooleanNormalizer DEFAULT = new BooleanNormalizer(DEFAULT_SIZE_LIMIT);
+
+    private final int sizeLimit;
+
+    private BooleanNormalizer(int sizeLimit) {
+        this.sizeLimit = sizeLimit;
+    }
+
+    /**
+     * Obtain a normalizer with the default size limit {@link BooleanNormalizer#DEFAULT_SIZE_LIMIT}.
+     * @return a normalizer with the default size limit
+     */
+    public static BooleanNormalizer getInstance() {
+        return DEFAULT;
     }
 
     /**
-     * Convert given predicate to disjunctive normal form, if necessary.
+     * Obtain a normalizer with the given size limit.
+     * @param sizeLimit a limit on the size of DNF that this normalizer will produce
+     * @return a normalizer with the given size limit
+     */
+    public static BooleanNormalizer withLimit(int sizeLimit) {
+        return new BooleanNormalizer(sizeLimit);
```

Comment:
If `sizeLimit == BooleanNormalizer.DEFAULT_LIMIT`, should this return `BooleanNormalizer.getInstance()`?

Reply:
I doubt this will make any difference ever, but sure, why not.